ingestion/src/main/java/feast/ingestion/transform/metrics/WriteFeatureValueMetricsDoFn.java

```diff
@@ -0,0 +1,312 @@
+
+/**
+ * WriteFeatureValueMetricsDoFn accepts key value of FeatureSetRef(str) to FeatureRow(List) and
+ * writes a histogram of the numerical values of each feature to StatsD.
+ *
+ * <p>The histogram of the numerical values is represented as the following in StatsD:
+ *
+ * <ul>
+ *   <li>gauge of feature_value_min
+ *   <li>gauge of feature_value_max
+ *   <li>gauge of feature_value_mean
+ *   <li>gauge of feature_value_percentile_50
+ *   <li>gauge of feature_value_percentile_90
+ *   <li>gauge of feature_value_percentile_95
+ * </ul>
+ *
+ * <p>StatsD timing/histogram metric type is not used since it does not support negative values.
+ */
+@AutoValue
+public abstract class WriteFeatureValueMetricsDoFn
+    extends DoFn<KV<String, Iterable<FeatureRow>>, Void> {
+
+  abstract String getStoreName();
+
+  abstract String getStatsdHost();
+
+  abstract int getStatsdPort();
+
+  static Builder newBuilder() {
+    return new AutoValue_WriteFeatureValueMetricsDoFn.Builder();
+  }
+
+  @AutoValue.Builder
+  abstract static class Builder {
+
+    abstract Builder setStoreName(String storeName);
+
+    abstract Builder setStatsdHost(String statsdHost);
+
+    abstract Builder setStatsdPort(int statsdPort);
+
+    abstract WriteFeatureValueMetricsDoFn build();
+  }
+
+  private static final Logger log =
+      org.slf4j.LoggerFactory.getLogger(WriteFeatureValueMetricsDoFn.class);
+  private StatsDClient statsDClient;
+  public static String GAUGE_NAME_FEATURE_VALUE_MIN = "feature_value_min";
+  public static String GAUGE_NAME_FEATURE_VALUE_MAX = "feature_value_max";
+  public static String GAUGE_NAME_FEATURE_VALUE_MEAN = "feature_value_mean";
+  public static String GAUGE_NAME_FEATURE_VALUE_PERCENTILE_50 = "feature_value_percentile_50";
+  public static String GAUGE_NAME_FEATURE_VALUE_PERCENTILE_90 = "feature_value_percentile_90";
+  public static String GAUGE_NAME_FEATURE_VALUE_PERCENTILE_95 = "feature_value_percentile_95";
+
+  @Setup
+  public void setup() {
+    // Note that exception may be thrown during StatsD client instantiation but no exception
+    // will be thrown when sending metrics (mimicking the UDP protocol behaviour).
+    // https://jar-download.com/artifacts/com.datadoghq/java-dogstatsd-client/2.1.1/documentation
+    // https://github.com/DataDog/java-dogstatsd-client#unix-domain-socket-support
+    try {
+      statsDClient = new NonBlockingStatsDClient(METRIC_PREFIX, getStatsdHost(), getStatsdPort());
+    } catch (StatsDClientException e) {
+      log.warn("StatsD client cannot be started: " + e.getMessage());
+    }
+  }
+
+  @Teardown
+  public void tearDown() {
+    if (statsDClient != null) {
+      statsDClient.close();
+    }
+  }
+
+  @ProcessElement
+  public void processElement(
+      ProcessContext context,
+      @Element KV<String, Iterable<FeatureRow>> featureSetRefToFeatureRows) {
+    if (statsDClient == null) {
+      return;
+    }
+
+    String featureSetRef = featureSetRefToFeatureRows.getKey();
```

Comment:
I am a bit concerned that we are re implementing this code over and over. Big room for inconsistencies to arise, but I think we can probably make a separate issue for that.

Reply:
Ok